### DISCUSSION TITLE
Missing curator settings for some indices

### DISCUSSION THREAD

**Discussion by abesinger** [Security-Onion-Solutions/securityonion#6078] [Unsupported Versions] [answered]
I'm not sure whether the problem I'm about to describe is technically a bug; it's an omission that I discovered the hard way -- by losing data. In short, there are a bunch of Curator "action" files in `/opt/so/saltstack/default/salt/curator/files/action` but no corresponding default settings in `/opt/so/saltstack/local/pillar/global.sls`  The result is that if you follow the documentation and change all curator settings in `global.sls` to keep indices open longer, some indices will still get closed after 30 days.  You have to _add_ entries to `global.sls` for those indices (and to be clear, I'm not talking about custom indices). That is very **non**-obvious from the documentation, in fact, looking at the curator action template file, I'd say that it's a complete omission.  I've provided a detailed narrative below of what I discovered, but here's my suggestion on what should get changed:

1. Ensure that for every Curator action file/index that SO installs, there are corresponding Curator entries in `global.sls`
2. And/or have a global default Curator settings in `global.sls` for `close` and `delete` instead of having a hardcoded default in the action file template, so that one can set minimum close and delete times for all indices. Then the per-index settings are only necessary when deviating from the default.
3. Make the documentation more thorough, and as I said in #6074, make it clear in the installation/configuration sections of the documentation that there are defaults that should be modified.  Losing logs is a big deal.
4. 30 days is a pretty short default for keeping logs. Consider a longer default.

Here's my specific example and how I lost data: As I mentioned in #6074, I noticed (and panicked at first) that a bunch of my indices were closed.  After reading the Curator section of the SO documentation (and working through the documentation errors that were since fixed in #5595), I changed all the "close" settings to a higher number (730, and yes I know that's huge but I've got a plan for that). All good, or so I thought. I could see that e.g. `so-filebeat-*` indices weren't getting closed. As I am not a Curator guru, and since SO generally manages settings through the pillars, I though I had covered everything.

Here's the rub:  I had just stared collecting logs from a Cisco ASA using the filebeat cisco module, which created indices named `so-cisco-*`.  Curator has action files `so-cisco-close.yml` and `so-cisco-delete.yml` with settings for 30 days closure.  Since there were no settings present in `global.sls` for `so-cisco`, they remained at 30 days, and Curator started closing the indices.

But wait, there's more... if the indices were closed, I could just reopen them, no big loss, right?  Well, what also happened is that my search nodes hit 80% utilization on /nsm, so Curator started looking for closed indices to delete, and found the `so-cisco` indices, and started deleting them.  Pair that with me going on vacation for a week, and several weeks of data got deleted.

Looking at one of the action file templates, it's clear that it was set up to accept settings from the pillar file.  The first line of `so-cisco-close.yml` has the following:
```
{%- set cur_close_days = salt['pillar.get']('elasticsearch:index_settings:so-cisco:close', 30) -%}
```
For those not familiar with jinja2 templates, that means translates to "set cur_close_days to the value in  `elasticsearch.index_settings.so-cisco.close`, or to 30 if it doesn't exist"

Had there been a corresponding entry in global.sls, I would have modified it from the get-go, and not had this problem.

Looking at the contents of `/opt/so/saltstack/default/salt/curator/files/action`, there are a action files for a _whole bunch_ of indices:
```
# ls /opt/so/saltstack/default/salt/curator/files/action
delete.yml                   so-gcp-delete.yml               so-osquery-close.yml
so-aws-close.yml             so-gcp-warm.yml                 so-osquery-delete.yml
so-aws-delete.yml            so-google_workspace-close.yml   so-osquery-warm.yml
so-aws-warm.yml              so-google_workspace-delete.yml  so-ossec-close.yml
so-azure-close.yml           so-google_workspace-warm.yml    so-ossec-delete.yml
so-azure-delete.yml          so-ids-close.yml                so-ossec-warm.yml
so-azure-warm.yml            so-ids-delete.yml               so-proofpoint-close.yml
so-barracuda-close.yml       so-ids-warm.yml                 so-proofpoint-delete.yml
so-barracuda-delete.yml      so-imperva-close.yml            so-proofpoint-warm.yml
so-barracuda-warm.yml        so-imperva-delete.yml           so-radware-close.yml
so-beats-close.yml           so-imperva-warm.yml             so-radware-delete.yml
so-beats-delete.yml          so-import-close.yml             so-radware-warm.yml
so-beats-warm.yml            so-import-delete.yml            so-redis-close.yml
so-bluecoat-close.yml        so-import-warm.yml              so-redis-delete.yml
so-bluecoat-delete.yml       so-infoblox-close.yml           so-redis-warm.yml
so-bluecoat-warm.yml         so-infoblox-delete.yml          so-snort-close.yml
so-cef-close.yml             so-infoblox-warm.yml            so-snort-delete.yml
so-cef-delete.yml            so-juniper-close.yml            so-snort-warm.yml
so-cef-warm.yml              so-juniper-delete.yml           so-snyk-close.yml
so-checkpoint-close.yml      so-juniper-warm.yml             so-snyk-delete.yml
so-checkpoint-delete.yml     so-kibana-close.yml             so-snyk-warm.yml
so-checkpoint-warm.yml       so-kibana-delete.yml            so-sonicwall-close.yml
so-cisco-close.yml           so-kibana-warm.yml              so-sonicwall-delete.yml
so-cisco-delete.yml          so-logstash-close.yml           so-sonicwall-warm.yml
so-cisco-warm.yml            so-logstash-delete.yml          so-sophos-close.yml
so-cyberark-close.yml        so-logstash-warm.yml            so-sophos-delete.yml
so-cyberark-delete.yml       so-microsoft-close.yml          so-sophos-warm.yml
so-cyberark-warm.yml         so-microsoft-delete.yml         so-strelka-close.yml
so-cylance-close.yml         so-microsoft-warm.yml           so-strelka-delete.yml
so-cylance-delete.yml        so-misp-close.yml               so-strelka-warm.yml
so-cylance-warm.yml          so-misp-delete.yml              so-syslog-close.yml
so-elasticsearch-close.yml   so-misp-warm.yml                so-syslog-delete.yml
so-elasticsearch-delete.yml  so-netflow-close.yml            so-syslog-warm.yml
so-elasticsearch-warm.yml    so-netflow-delete.yml           so-tomcat-close.yml
so-f5-close.yml              so-netflow-warm.yml             so-tomcat-delete.yml
so-f5-delete.yml             so-netscout-close.yml           so-tomcat-warm.yml
so-f5-warm.yml               so-netscout-delete.yml          so-zeek-close.yml
so-firewall-close.yml        so-netscout-warm.yml            so-zeek-delete.yml
so-firewall-delete.yml       so-o365-close.yml               so-zeek-warm.yml
so-firewall-warm.yml         so-o365-delete.yml              so-zscaler-close.yml
so-fortinet-close.yml        so-o365-warm.yml                so-zscaler-delete.yml
so-fortinet-delete.yml       so-okta-close.yml               so-zscaler-warm.yml
so-fortinet-warm.yml         so-okta.delete.yml              warm.yml
so-gcp-close.yml             so-okta-warm.yml
```

**Comment by dougburks** [comment, marked as answer]
> 3\. Make the documentation more thorough, and as I said in [Index closing/deletion and documentation #6074](https://github.com/Security-Onion-Solutions/securityonion/discussions/6074), make it clear in the installation/configuration sections of the documentation that there are defaults that should be modified.

As mentioned at https://github.com/Security-Onion-Solutions/securityonion/discussions/6074, I've updated https://docs.securityonion.net/en/2.3/curator.html and https://docs.securityonion.net/en/2.3/elasticsearch.html and also updated https://docs.securityonion.net/en/2.3/post-installation.html to include links to both of them.

> 4\. 30 days is a pretty short default for keeping logs.

We've tried very hard to come up with sensible defaults for most users. 30 days is a sensible default for open indices since having more than that can cause performance issues for some folks. From https://docs.securityonion.net/en/2.3/elasticsearch.html:
_The more indices are open, the more heap is required. Having too many open indices can lead to performance issues. There are many factors that determine the number of days you can have in an open state, so this is a good setting to adjust specific to your environment._

> if the indices were closed, I could just reopen them, no big loss, right? Well, what also happened is that my search nodes hit 80% utilization on /nsm

I think this is really the biggest contributing factor to your situation. Depending on your configuration, if your system reached the `log_size_limit` threshold, then curator may have started deleting older indices (closed and even open indices) to prevent your disk from filling. This is a form of self-protection as filling the disk would create other issues. Seems like maybe you need to adjust `log_size_limit` and possibly add more storage to meet your retention goals.

**Reply by abesinger** [reply]
Thanks for the feedback. Rereading this a second time, I noticed you said " if your system reached the `log_size_limit` threshold, curator may have started deleting older indices (closed and even open indices)".  Did you mean that Curator can delete indices independent of `unit_count`, or did you mean that it will delete open indices older than `unit_count`?  If the former, I can't find anything in the documentation that indicates it will do that.

**Reply by abesinger** [reply]
Regarding your other comments, yes, I'm aware of heap and number of indices, etc. have been tweaking settings, and am making plans to add more resources. What went wrong here is indices *younger* than other indices got deleted, because I thought I had addressed all the aging settings in `global.sls,` but hadn't because of missing entries.

Since you pointed out doing PRs in #6074, I'll look into doing a PR to update global.sls and maybe add a configurable global minimum age.  I also notice that there's no action file for so-panw (logs from the Palo Alto filebeat module), so I'll look into adding that too.

**Reply by dougburks** [reply]
> Did you mean that Curator can delete indices independent of `unit_count`, or did you mean that it will delete open indices older than `unit_count`?

Yes, for most configurations, Curator can delete indices independent of the individual index `unit_count` values. This is a safety precaution to avoid filing the disk which would create other problems.

Take a look at `/opt/so/conf/curator/action/` and see if you have a `delete.yml` file. This will be based on https://github.com/Security-Onion-Solutions/securityonion/blob/master/salt/curator/files/action/delete.yml but with the appropriate `log_size_limit` for your box.

> I can't find anything in the documentation that indicates it will do that.

From https://docs.securityonion.net/en/2.3/curator.html#configuration:
```
As your disk reaches capacity, Curator starts deleting old indices to prevent your disk from filling up. To change the limit, modify log_size_limit in /opt/so/saltstack/local/pillar/minions/$SENSORNAME_$ROLE.sls.
```

**Comment by abesinger** [comment]
@dougburks: thanks for the reply, If you're in a tl;dr mood, skip to the bottom for a few questions (and one possible bug), the rest remains for anyone who's interested in this thread.

I had looked at delete.yml (and all the other action files; I try to do my homework before posting, not always successfully). I had also read the Curator documentation trying to understand where only closed indices get deleted vs any index.  Somewhere in the process I thought I read that only open indices older than `elasticsearch.index_settings.so-*.delete` would ever get deleted.   But looking at just `delete.yml` and the Curator documentation on the `delete` action and the `age` filter, there's nothing that says that it applies to only closed indices.

But I see now what contributed to my misunderstanding. I had also looked at the `so-*` shell scripts in `/sbin`.  There are three scripts that manage deletion.  I had looked at them before but not quite grokked how they related to each other. `so-curator-closed-delete`, `so-curator-closed-delete-delete`, and `so-curator-delete`. 

`so-curator-closed-delete` is just a wrapper around `so-curator-closed-delete-delete`.  The latter incrementally deletes closed indices (only) until disk space gets below `log_size_limit` (or there are no more closed indices). It uses the Elastic API directly. Neither of those scripts actually invoke Curator.

`so-curator-delete` invokes Curator with `delete.yml`, which as I understand now will delete *any* index, oldest first, to get below `log_size_limit`.

`so-curator-delete` and `so-curator-closed-delete` get invoked by cron every 5 minutes.
 
Having seen `so-curator-closed-delete-delete` first, I think I made an assumption that such a script was necessary to actually get indices deleted, and that Curator relied on such a script to be available.  I understand now that Curator does it's own index removal, and that `so-curator-closed-delete*` actually have nothing to do with Curator, regardless of how they are named.

I also see that there are action files named `/opt/so/conf/curator/so-*-delete.yml` that get invoked by `/sbin/so-curator-cluster-delete`, but I don't see anything that in turn invokes that shell script.

So I think I understand now what's going on.  But I'm curious about a few things:
* Why is `so-curator-closed-delete-delete` not using Curator?
* The cron invocation of `so-curator-delete` and `so-curator-closed-delete` does nothing to ensure that they are run in any particular order (in fact, they'll likely get executed in parallel).  Is that intentional? I'd think you'd always want to run `so-curator-closed-delete` first, no? 
* Is `so-curator-cluster-delete` ever invoked by anything?

**Reply by dougburks** [reply]
> * Why is `so-curator-closed-delete-delete` not using Curator?

I've added the following to https://docs.securityonion.net/en/2.3/elasticsearch.html:
so-curator-closed-delete does not use Curator because it cannot calculate disk space used by closed indices. For more information, see https://www.elastic.co/guide/en/elasticsearch/client/curator/current/filtertype_space.html.

> * The cron invocation of `so-curator-delete` and `so-curator-closed-delete` does nothing to ensure that they are run in any particular order (in fact, they'll likely get executed in parallel).  Is that intentional? I'd think you'd always want to run `so-curator-closed-delete` first, no?

EDITED 2021-11-10 The nuances of Curator meant that it took a few rounds of internal discussion, sleep, and caffeine to remember all the details of why this works the way it does. I've added 2 new sections to the Elasticsearch doc page:
https://docs.securityonion.net/en/2.3/elasticsearch.html#closing-indices
https://docs.securityonion.net/en/2.3/elasticsearch.html#deleting-indices

To answer your question, running these jobs in parallel shouldn't be an issue since Curator cannot calculate disk space used by closed indices. If we have both open and closed indices, we may be at log_size_limit but curator's delete.yml is going to see index disk usage at a value lower than log_size_limit and so it shouldn't delete any open indices.

Example:

Suppose our log_size_limit is 1TB and we have 30 days of open indices and 300 days of closed indices. We reach log_size_limit and the cron jobs fire in parallel. Curator's delete.yml will check disk space used but it will see that disk space is at maybe 100GB so it thinks we haven't reached log_size_limit and does not delete anything. so-curator-closed-delete gets a more accurate view of disk space used, sees that we have indeed reached log_size_limit, and so it deletes closed indices until we get lower than log_size_limit.

> * Is `so-curator-cluster-delete` ever invoked by anything?

Yes, but only if you're using Elastic clustering:
https://docs.securityonion.net/en/2.3/elasticsearch.html#elastic-clustering

**Reply by abesinger** [reply]
> To answer your question, running these jobs in parallel shouldn't be an issue since Curator cannot calculate disk space used by closed indices. If we have both open and closed indices, we may be at log_size_limit but curator's delete.yml is going to see index disk usage at a value lower than log_size_limit and so it shouldn't delete any open indices.

Ah, I get it now.  Curator does not look at the partition size, only the amount of space occupied by open indices. Great documentation addition too, thanks!

BTW, I am aware that Elastic doesn't report the size of close indices,y.  Honestly, seems like it wouldn't be that hard for Elastic to do so; I believe indices are kept in files separate from each other, so it wouldn't be difficult or a lot of data to manage, but maybe there's a nuance to how Elastic stores indices that I'm not aware of.  Regardless, that's an issue for Elastic, not for you.

> Yes, but only if you're using Elastic clustering: https://docs.securityonion.net/en/2.3/elasticsearch.html#elastic-clustering

That's also helpful to know.  I'm looking into doing a rebuild (technically, new build + migration) using Elastic Clustering.
